File: python-algorithm/ds_1_stack_queue/L150-m.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/2 8:40
# @Author  : Heng Li
# @File    : L150-m.py
# @Software: PyCharm

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stk = []
        for token in tokens:
            if token not in ['+', '-', '*', '/']:
                stk.append(token)
            else:
                tmp_1 = int(stk.pop(-1))
                tmp_2 = int(stk.pop(-1))
                if token == '+':
                    stk.append(tmp_2 + tmp_1)
                elif token == '-':
                    stk.append(tmp_2 - tmp_1)
                elif token == '*':
                    stk.append(tmp_2 * tmp_1)
                elif token == '/':
                    # 题目要求保留整数除法的整数部分，因此不能用python 的'//'向下整除在负数时候回小1
                    stk.append(tmp_2 / tmp_1)
        return int(stk.pop(-1))

File: python-algorithm/ds_1_stack_queue/L150-m.cpp

#include <iostream>
#include <stack>
#include <vector>
#include <sstream>

//<sstream>库定义了三种类：istringstream、ostringstream和stringstream，分别用来进行流的输入、输出和输入输出操作。

using namespace std;
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> stk;
        for(int i = 0 ;i < tokens.size(); i++){
            if(tokens[i] == "+"){
                int a = stk.top();
                stk.pop();
                int b = stk.top();
                stk.pop();
                stk.push(b+a);
            }
            else if(tokens[i] == "-"){
                int a = stk.top();
                stk.pop();
                int b = stk.top();
                stk.pop();
                stk.push(b-a);
            }
            else if(tokens[i] == "*"){
                int a = stk.top();
                stk.pop();
                int b = stk.top();
                stk.pop();
                stk.push(b*a);
            }
            else if(tokens[i] == "/"){
                int a = stk.top();
                stk.pop();
                int b = stk.top();
                stk.pop();
                stk.push(b/a);
            }
            else
                stk.push(string2Int(tokens[i]));
        }
        return stk.top();
    }

    //数字字符串转数字
    int string2Int(string x ){
        int a  ;
        stringstream ss(x);
        ss >> a;
        return a;
    }
};